File: oop.js
// Class 1: Computer
class Computer {
  constructor(name, RAM, SSD, OS) {
    this.name = name;
    this.RAM = RAM;
    this.SSD = SSD;
    this.OS = OS;
  }

  specs() {
    return `${this.name}, has ${this.RAM}-GB RAM on a ${this.SSD} GB SSD, running ${this.OS}.`;
  }

  upgrade() {
    this.RAM++;
    return `${this.name} is now ${this.RAM} GB`;
  }

  changeOS(newOS) {
    this.OS = newOS;
    return `${this.name} has a new OS: ${this.OS}`;
  }

  update() {
    this.OS = "updated";
    return `${this.name} has updated from previous version.`;
  }
}

// Class 2: Phone
class Phone {
  constructor(make, model, year, color) {
    this.make = make;
    this.model = model;
    this.year = year;
    this.color = color;
  }

  setup() {
    return `Setting up the ${this.year} ${this.color} ${this.make} ${this.model}.`;
  }

  run() {
    return `running the ${this.make} ${this.model}.`;
  }

  updates() {
    return `Updateing the ${this.make} ${this.model}.`;
  }

  customize() {
    return `applying the customizations of the ${this.make} ${this.model}.`;
  }
}


// Subclass Iphone
class Iphone extends Phone {
    constructor(make, model, year, color, titanium) {
      super(make, model, year, color); 
      this.titanium = titanium; 
    }

    actionButton() {
        return ` ${this.make} ${this.model} has new action button.`;
      }
  

  }
  
 

// Class 3: Book
class Book {
  constructor(title, author, genre, pageCount) {
    this.title = title;
    this.author = author;
    this.genre = genre;
    this.pageCount = pageCount;
  }

  getInfo() {
    return `${this.title} by ${this.author} (${this.genre}) - ${this.pageCount} pages`;
  }

  read() {
    return `Reading ${this.title}...`;
  }

  recommend() {
    return `I recommend reading ${this.title} by ${this.author}.`;
  }

  bookmark(pageNumber) {
    return `Bookmarking page ${pageNumber} of ${this.title}.`;
  }
}


// Class 4: Movie
class Movie {
    constructor(title, director, genre, runTime, timeFrame) {
      this.title = title;
      this.director = director;
      this.genre = genre;
      this.runTime = runTime;
      this.timeFrame = timeFrame;
    }
  
    getInfo() {
      return `${this.title} by ${this.director} (${this.genre}) - ${this.runTime} pages`;
    }
  
    watch() {
      return `watching ${this.title}...`;
    }
  
    recommend() {
      return `I recommend watching ${this.title} by ${this.director}.`;
    }
  
    bookmark() {
      return `Bookmarking the time frame ${this.timeFrame} of ${this.title}.`;
    }
  }


// objects of the classes
const computer1 = new Computer("Dell", 32, 500, "chromeOS");
const phone1 = new Phone("Google Pixel", 8, 2023, "White");
const book1 = new Book("The Lord of the Rings", "J.R.Tolkins", "Fiction", 500);
const movie1 = new Movie("The Lord of the Rings", "Peter Jackson", "Fiction", "2 hours","2:33");
const myIphone = new Iphone("Apple", "iPhone 13", 2023, "Silver", true);

console.log(computer1.specs());
console.log(phone1.setup());
console.log(book1.getInfo());
console.log(movie1.bookmark());
console.log(myIphone.titanium); 

File: oop.py
# Class 1: Computer
class Computer:
    def __init__(self, name, RAM, SSD, OS):
        self.name = name
        self.RAM = RAM
        self.SSD = SSD
        self.OS = OS

    def specs(self):
        return f"{self.name}, has {self.RAM}-GB RAM on a {self.SSD} GB SSD, running {self.OS}."

    def upgrade(self):
        self.RAM += 1
        return f"{self.name} is now {self.RAM} GB"

    def changeOS(self, newOS):
        self.OS = newOS
        return f"{self.name} has a new OS: {self.OS}"

    def update(self):
        self.OS = "updated"
        return f"{self.name} has updated from the previous version."

# Class 2: Phone
class Phone:
    def __init__(self, make, model, year, color):
        self.make = make
        self.model = model
        self.year = year
        self.color = color

    def setup(self):
        return f"Setting up the {self.year} {self.color} {self.make} {self.model}."

    def run(self):
        return f"Running the {self.make} {self.model}."

    def updates(self):
        return f"Updating the {self.make} {self.model}."

    def customize(self):
        return f"Applying customizations of the {self.make} {self.model}."


# Subclass Iphone
class Iphone(Phone):
    def __init__(self, make, model, year, color, titanium):
        super().__init__(make, model, year, color)
        self.titanium = titanium

    def action_button(self):
        return f"{self.make} {self.model} has a new action button."




# Class 3: Book
class Book:
    def __init__(self, title, author, genre, pageCount):
        self.title = title
        self.author = author
        self.genre = genre
        self.pageCount = pageCount

    def getInfo(self):
        return f"{self.title} by {self.author} ({self.genre}) - {self.pageCount} pages"

    def read(self):
        return f"Reading {self.title}..."

    def recommend(self):
        return f"I recommend reading {self.title} by {self.author}."

    def bookmark(self, pageNumber):
        return f"Bookmarking page {pageNumber} of {self.title}."

# Class 4: Movie
class Movie:
    def __init__(self, title, director, genre, runTime, timeFrame):
        self.title = title
        self.director = director
        self.genre = genre
        self.runTime = runTime
        self.timeFrame = timeFrame

    def getInfo(self):
        return f"{self.title} by {self.director} ({self.genre}) - {self.runTime} minutes"

    def watch(self):
        return f"Watching {self.title}..."

    def recommend(self):
        return f"I recommend watching {self.title} by {self.director}."

    def bookmark(self):
        return f"Bookmarking the time frame {self.timeFrame} of {self.title}."

# objects of the classes
computer1 = Computer("Dell", 32, 500, "chromeOS")
phone1 = Phone("Google Pixel", 8, 2023, "White")
book1 = Book("The Lord of the Rings", "J.R.R. Tolkien", "Fiction", 500)
movie1 = Movie("The Lord of the Rings", "Peter Jackson", "Fiction", "2 hours", "2:33")
my_iphone = Iphone("Apple", "iPhone 13", 2023, "Silver", True)

print(computer1.specs())
print(phone1.setup())
print(book1.getInfo())
print(movie1.bookmark())
print(my_iphone.action_button())
